lib/Doctrine/ORM/Persisters/ManyToManyPersister.php

```diff
@@ -37,130 +37,288 @@ class ManyToManyPersister extends AbstractCollectionPersister
 {
     /**
      * {@inheritdoc}
-     *
-     * @override
      */
-    protected function getDeleteRowSQL(PersistentCollection $coll)
+    public function delete(PersistentCollection $coll)
     {
-        $columns    = array();
-        $mapping    = $coll->getMapping();
-        $class      = $this->em->getClassMetadata(get_class($coll->getOwner()));
-        $tableName  = $this->quoteStrategy->getJoinTableName($mapping, $class, $this->platform);
+        $mapping = $coll->getMapping();
 
-        foreach ($mapping['joinTable']['joinColumns'] as $joinColumn) {
-            $columns[] = $this->quoteStrategy->getJoinColumnName($joinColumn, $class, $this->platform);
+        if ( ! $mapping['isOwningSide']) {
+            return; // ignore inverse side
         }
 
-        foreach ($mapping['joinTable']['inverseJoinColumns'] as $joinColumn) {
-            $columns[] = $this->quoteStrategy->getJoinColumnName($joinColumn, $class, $this->platform);
+        $this->conn->executeUpdate($this->getDeleteSQL($coll), $this->getDeleteSQLParameters($coll));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update(PersistentCollection $coll)
+    {
+        $mapping = $coll->getMapping();
+
+        if ( ! $mapping['isOwningSide']) {
+            return; // ignore inverse side
         }
 
-        return 'DELETE FROM ' . $tableName
-            . ' WHERE ' . implode(' = ? AND ', $columns) . ' = ?';
+        $diff   = $coll->getDeleteDiff();
+        $sql    = $this->getDeleteRowSQL($coll);
+
+        foreach ($diff as $element) {
+            $this->conn->executeUpdate($sql, $this->getDeleteRowSQLParameters($coll, $element));
+        }
+
+        $diff   = $coll->getInsertDiff();
```

Comment:
Shouldn't we handle insert operations before? This seems to clash with the [UnitOfWork semantics](https://github.com/doctrine/doctrine2/blob/b9c0868f08666a61962f9581a415397118bf99fe/lib/Doctrine/ORM/UnitOfWork.php#L357-L389) (inserts first)